api/websocket/server/relay.go

```diff
@@ -124,11 +125,44 @@ func (ws *WsServer) sendInboundRelayMessage(relayMessage *pb.Relay) {
 				sigChainLen: int(relayMessage.SigChainLen),
 			})
 		}
-	} else {
+		if time.Duration(relayMessage.MaxHoldingSeconds) > pongTimeout/time.Second {
+			ok := ws.messageDeliveredCache.Push(relayMessage)
+			if !ok {
+				log.Warningf("MessageDeliveredCache full, discarding messages.")
+			}
+		}
+	} else if relayMessage.MaxHoldingSeconds > 0 {
 		ws.messageBuffer.AddMessage(clientID, relayMessage)
 	}
 }
 
+func (ws *WsServer) startCheckingLostMessages() {
+	for {
+		v, ok := ws.messageDeliveredCache.Pop()
```

Comment:
Where to push again if the popped msg still can't deliver?

Reply:
When msg from messageBuffer failed to deliver, it will be put back into buffer again